Storyshots: Fix autotitle

tmeasday opened the pull request:
Issue: #16048 

## What I did

- Add a storyshots setup for `react-ts`
- Set `global.STORIES` in storyshots

This works (for `react-ts`) in v7 mode because we never set `global.FEATURES` in Storyshots, so it runs in v6 mode anyway, for better or worse. AFAIK this isn't a problem but isn't a sustainable strategy (maybe OK for now?)

## What still needs to happen

1. This PR relies on this PR: https://github.com/smrq/babel-plugin-require-context-hook/pull/14

2. We need to figure out a strategy for story sorting (perhaps the one difference between v7 and v6 stores?)

## How to test

- [x] Is this testable with Jest or Chromatic screenshots?

Hopefully these tests are getting run (they should fail right now).

tmeasday committed:
Add a storyshots test for `react-ts`

tmeasday committed:
Set `STORIES` in storyshots to ensure autoTitle works

tmeasday commented on lib/client-api/src/StoryStoreFacade.ts:
```diff
@@ -23,8 +23,6 @@ import {
   StoryIndexEntry,
 } from '@storybook/store';
 
-const { STORIES = [] } = global;
```
We can't read this right away because it doesn't get set until after `require` in storyshots 🤮

ndelangen replied:
tbf I think this is actually an improvement.. when accessing things from the global scope, you want to access them JIT, to prevent memory leaks?

If global.STORIES changes over time, this module keeps a reference around of the first one for ever... the new code doesn't.

tmeasday replied:
I guess so, although why would it change over time? (Apart from this reason ;) )

shilman commented:
- [ ] Copy the feature flags
- [ ] Storyshots won't work in v7 mode for now
- [ ] Add workaround in MIGRATION.md

tmeasday committed:
Pass `FEATURES` into storyshots and read JIT in preview

tmeasday committed:
Add migration note about story store v7 and apply it in react-ts

tmeasday committed:
Ensure check works in node too

tmeasday commented:
OK done @shilman !

shilman reviewed (changes requested):
failing tests 😭

tmeasday committed:
Use sensible defaults when `FEATURES` or `STORIES` are unset

tmeasday commented:
So in this failing test:

```
FAIL addons/storyshots/storyshots-core/stories/storyshot.configFunc.test.js
662
  ● Test suite failed to run
663

664
    ENOENT: no such file or directory, lstat '/home/runner/work/storybook/storybook/.storybook'
665

666
      44 | function getConfigPathParts(input) {
667
      45 |     var configDir = path_1.default.resolve(input);
668
    > 46 |     if (fs_1.default.lstatSync(configDir).isDirectory()) {
669
         |                      ^
670
      47 |         var output = {};
671
      48 |         var preview = exports.getPreviewFile(configDir);
672
      49 |         var main = exports.getMainFile(configDir);
673

674
      at getConfigPathParts (addons/storyshots/storyshots-core/dist/ts3.9/frameworks/configure.js:46:22)
675
      at Object.configure [as default] (addons/storyshots/storyshots-core/dist/ts3.9/frameworks/configure.js:76:14)
676
      at Object.load (addons/storyshots/storyshots-core/dist/ts3.9/frameworks/react/loader.js:24:5)
677
      at Object.loadFramework [as default] (addons/storyshots/storyshots-core/dist/ts3.9/frameworks/frameworkLoader.js:26:19)
678
      at testStorySnapshots (addons/storyshots/storyshots-core/dist/ts3.9/api/index.js:42:14)
679
      at Object.<anonymous> (addons/storyshots/storyshots-core/stories/storyshot.configFunc.test.js:34:1)
```

The issue is that in this example we are providing a custom `config` function. Previously if we did that, we didn't try and read `main.js`.

In this PR we always read `main.js` so we can set `STORIES` and `FEATURES` correctly. However in cases like this one, where the config dir isn't set properly, this can be a breaking change.

WDYT @shilman? We could just not set `STORIES` and `FEATURES` if `config` is set. This is sort of wrong for `FEATURES` however.

shilman commented:
> WDYT @shilman? We could just not set STORIES and FEATURES if config is set. This is sort of wrong for FEATURES at least however.

Seems like a reasonable hack to me

tmeasday committed:
Don't set FEATURES if `config` is supplied in storyshots

(The only way to implement it would involve a breaking change, see: https://github.com/storybookjs/storybook/pull/16568#issuecomment-962495996)

tmeasday committed:
Merge remote-tracking branch 'origin/next' into 16048-v7-storyshots

tmeasday committed:
Update to our version of babel rc hook

shilman committed:
Merge branch 'next' into 16048-v7-storyshots

shilman committed:
Upgrade @storybook/babel-plugin-require-context-hook

shilman reviewed (approved):
Updated. Ready to merge I think! 🤔
